### PR TITLE
Support arm64 in docker driver/integration tests

### PR DESCRIPTION
**Pull request by ilya-zuyev**
This PR adds fixes to run integration tests on linux/arm64 platform

- docker driver implementation changed to use multi-arch version of kicbase
- some integration tests are disabled

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -654,7 +654,7 @@ ifndef AUTOPUSH
 	$(call user_confirm, 'Are you sure you want to push: $(KIC_BASE_IMAGE_GH) & $(KIC_BASE_IMAGE_GCR) & $(KIC_BASE_IMAGE_HUB) ?')
 	$(MAKE) push-kic-base-image AUTOPUSH=true
 else
-	$(MAKE) push-kic-base-image-gcr push-kic-base-image-hub push-kic-base-image-gh
```
Comment (medyagh) [inline Makefile:657]: revert or make unreleated changes in seprate PR

File: pkg/minikube/bootstrapper/images/images.go [hunk 2]
```diff
@@ -121,10 +121,7 @@ func etcd(v semver.Version, mirror string) string {
 
 // archTag returns a CPU architecture suffix for images
 func archTag(hasTag bool) string {
```
Comment (medyagh) [inline pkg/minikube/bootstrapper/images/images.go:123]: lets ensure this doesnt break the old kubernetes version as seen in this part of the code
https://github.com/medyagh/minikube/blob/75cd17801287339a00efe44fbd451b4dbfadef86/pkg/minikube/bootstrapper/images/images.go#L119

File: pkg/minikube/registry/drvs/docker/docker.go [hunk 3]
```diff
@@ -83,7 +83,7 @@ func configure(cc config.ClusterConfig, n config.Node) (interface{}, error) {
 }
 
 func status() registry.State {
-	if runtime.GOARCH != "amd64" {
+	if runtime.GOARCH != "amd64" && runtime.GOARCH != "arm64" {
```
Comment (medyagh) [inline pkg/minikube/registry/drvs/docker/docker.go:86]: lets add plain "arm" in a follow up PR

File: test/integration/aab_offline_test.go [hunk 4]
```diff
@@ -29,20 +29,24 @@ import (
...
-			runtime := runtime
-			t.Run(runtime, func(t *testing.T) {
+		for _, rt := range []string{"docker", "crio", "containerd"} {
+			rt := rt
+			t.Run(rt, func(t *testing.T) {
 				MaybeParallel(t)
 
-				if runtime != "docker" && NoneDriver() {
+				if rt != "docker" && Arm64Platform() {
+					t.Skipf("skipping %s - only docker runtime supported on arm64", t.Name())
```
Comment (medyagh) [inline test/integration/aab_offline_test.go:38]: if it passes the test, lets not skip it

**Review by medyagh — CHANGES_REQUESTED** [review]

File: test/integration/aab_offline_test.go (hunk 4 above)
Comment (tstromberg) [reply test/integration/aab_offline_test.go:38]: I think this is doing the right thing: skip arm64 tests if we aren't on Docker.

File: test/integration/addons_test.go [hunk 5]
```diff
@@ -41,6 +41,11 @@ import (
 
 // TestAddons tests addons that require no special environment -- in parallel
 func TestAddons(t *testing.T) {
+
+	if !Arm64Platform() {
+		t.Skipf("Skip helm addon test for arm64")
```
Comment (tstromberg) [inline test/integration/addons_test.go:46]: Can you add a GitHub issue to track fixing this? That way you can put the URL in the skip message.

File: test/integration/main_test.go [hunk 6]
```diff
@@ -135,6 +135,11 @@ func GithubActionRunner() bool {
 	return os.Getenv("GITHUB_ACTIONS") == "true"
 }
 
+// Arm64Platform returns true if running on arm64/* platform
+func Arm64Platform() bool {
```
Comment (tstromberg) [inline test/integration/main_test.go:139]: I feel like this helper can be removed. It saves zero lines of code and abstracts away an important distinction.

File: test/integration/functional_test.go [hunk 7]
```diff
@@ -591,7 +606,13 @@ func validateCacheCmd(ctx context.Context, t *testing.T, profile string) {
 			if err != nil {
 				t.Errorf("failed to get images by %q ssh %v", rr.Command(), err)
 			}
-			if !strings.Contains(rr.Output(), "0184c1613d929") {
+			var pauseID string
+			if Arm64Platform() {
```
Comment (tstromberg) [inline test/integration/functional_test.go:610]: Can you make this a switch statement or map containing hashes for different architectures? I feel like we shouldn't hard-code yet another architecture here, when there will certainly be others in the near-term support horizon, such as riscv.

File: test/integration/skaffold_test.go [hunk 8]
```diff
@@ -38,6 +38,9 @@ func TestSkaffold(t *testing.T) {
 	if NoneDriver() {
 		t.Skip("none driver doesn't support `minikube docker-env`; skaffold depends on this command")
 	}
+	if Arm64Platform() {
+		t.Skip("none driver doesn't support `minikube docker-env`; skaffold depends on this command")
```
Comment (tstromberg) [inline test/integration/skaffold_test.go:42]: This doesn't make sense to me.

**Review by tstromberg — CHANGES_REQUESTED** [review]

File: test/integration/addons_test.go (hunk 5 above)
Comment (ilya-zuyev) [reply test/integration/addons_test.go:46]: https://github.com/kubernetes/minikube/issues/10144

File: test/integration/main_test.go (hunk 6 above)
Comment (ilya-zuyev) [reply test/integration/main_test.go:139]: For me this helper is a kind of "bookmark function". It's used only in integration tests and helps to find all places where we have exceptions for arm64.

File: Makefile (hunk 1 above)
Comment (ilya-zuyev) [reply Makefile]: done

File: pkg/minikube/registry/drvs/docker/docker.go (hunk 3 above)
Comment (ilya-zuyev) [reply pkg/minikube/registry/drvs/docker/docker.go]: It's TODO. For now arm32 is not supported by kicbase Dockerfile.

File: test/integration/functional_test.go (hunk 7 above)
Comment (ilya-zuyev) [reply test/integration/functional_test.go]: Done

File: test/integration/skaffold_test.go (hunk 8 above)
Comment (ilya-zuyev) [reply test/integration/skaffold_test.go]: copy-pasted message, my bad :( 
Removed.

File: pkg/minikube/bootstrapper/images/images.go [hunk 9]
```diff
@@ -120,8 +121,8 @@ func etcd(v semver.Version, mirror string) string {
 }
 
 // archTag returns a CPU architecture suffix for images
-func archTag(hasTag bool) string {
-	if runtime.GOARCH == "amd64" && !hasTag {
+func archTag(needsArchSuffix bool) string {
```
Comment (medyagh) [inline pkg/minikube/bootstrapper/images/images.go:124]: how about adding unittest for this func to be more clear what we need?

File: test/integration/aaa_download_only_test.go [hunk 10]
```diff
@@ -194,13 +194,15 @@ func TestDownloadOnlyKic(t *testing.T) {
...
 	}
-	// Make sure it has the correct checksum
-	checksum := md5.Sum(contents)
-	remoteChecksum, err := ioutil.ReadFile(download.PreloadChecksumPath(constants.DefaultKubernetesVersion, cRuntime))
-	if err != nil {
-		t.Errorf("failed to read checksum file %q : %v", download.PreloadChecksumPath(constants.DefaultKubernetesVersion, cRuntime), err)
-	}
-	if string(remoteChecksum) != string(checksum[:]) {
-		t.Errorf("failed to verify checksum. checksum of %q does not match remote checksum (%q != %q)", tarball, string(remoteChecksum), string(checksum[:]))
+	if !arm64Platform() {
```
Comment (medyagh) [inline test/integration/aaa_download_only_test.go:197]: let use Skip for this ,so it will show up in the Gopogh Skip tests
```
if arm64Platform() { 
   t.Skip(....)
}
```

File: test/integration/addons_test.go (hunk 5 above)
Comment (medyagh) [reply test/integration/addons_test.go:46]: lets add the issue in the Skip Message.

simmilar to other Skips, lets add a github issue and Write the Issue in the Skip Message

File: test/integration/functional_test.go [hunk 11]
```diff
@@ -361,6 +361,23 @@ func validateExtraConfig(ctx context.Context, t *testing.T, profile string) {
 
 }
 
+func imageID(image string) string {
```
Comment (medyagh) [inline test/integration/functional_test.go:364]: add a comment what this func is for and why...

File: test/integration/functional_test.go [hunk 12]
```diff
@@ -493,7 +510,22 @@ func validateDryRun(ctx context.Context, t *testing.T, profile string) {
 	defer cancel()
 
 	// Too little memory!
-	startArgs := append([]string{"start", "-p", profile, "--dry-run", "--memory", "250MB", "--alsologtostderr"}, StartArgs()...)
+	var startArgs []string
+	orgStartArgs := StartArgs()
+
+	for i := 0; i < len(orgStartArgs); i++ {
```
Comment (medyagh) [inline test/integration/functional_test.go:516]: can u add a comment what this code block is doing ?
I rather integration code read be super simple so when we debug a test not have to read too carefuly to know what code is doing

File: test/integration/functional_test.go [hunk 13]
```diff
@@ -783,7 +815,14 @@ func validateServiceCmd(ctx context.Context, t *testing.T, profile string) {
...
 
-	rr, err := Run(t, exec.CommandContext(ctx, "kubectl", "--context", profile, "create", "deployment", "hello-node", "--image=k8s.gcr.io/echoserver:1.4"))
+	var echoServerArg string
+	if arm64Platform() {
+		echoServerArg = "k8s.gcr.io/echoserver-arm:1.8"
+	} else {
+		echoServerArg = "k8s.gcr.io/echoserver:1.10"
+	}
+
+	rr, err := Run(t, exec.CommandContext(ctx, "kubectl", "--context", profile, "create", "deployment", "hello-node", "--image="+echoServerArg))
```
Comment (medyagh) [inline test/integration/functional_test.go:825]: how about we have to Blocks that hve the full command for each  instead of multiple ifs
```

if arm64Platform() {
// add comment why it is different
rr, err := Run(t, exec.CommandContext(ctx, "kubectl", "--context", profile, "create", "deployment", "hello-node", "--image=k8s.gcr.io/echoserver-arm:1.8"))
} else {
rr, err := Run(t, exec.CommandContext(ctx, "kubectl", "--context", profile, "create", "deployment", "hello-node", "--image=k8s.gcr.io/echoserver:1.10"))

```

and then get rid of all before logic so code be more simple to read

**Review by medyagh — CHANGES_REQUESTED** [review]

File: test/integration/main_test.go (hunk 6 above)
Comment (ilya-zuyev) [reply test/integration/main_test.go]: fixed

File: test/integration/functional_test.go [hunk 14]
```diff
@@ -955,6 +994,11 @@ func validateSSHCmd(ctx context.Context, t *testing.T, profile string) {
 
 // validateMySQL validates a minimalist MySQL deployment
 func validateMySQL(ctx context.Context, t *testing.T, profile string) {
+	if Arm64Platform() {
+		t.Skip("arm64 is not supported by mysql. skip the test")
+		return
```
[the remark this answers is not in the document]
Comment (ilya-zuyev) [reply test/integration/functional_test.go]: Removed

File: pkg/minikube/bootstrapper/images/images.go (hunk 9 above)
Comment (ilya-zuyev) [reply pkg/minikube/bootstrapper/images/images.go:124]: done

File: test/integration/aaa_download_only_test.go (hunk 10 above)
Comment (ilya-zuyev) [reply test/integration/aaa_download_only_test.go]: done

File: test/integration/functional_test.go (hunk 12 above)
Comment (ilya-zuyev) [reply test/integration/functional_test.go:523]: done

File: test/integration/functional_test.go (hunk 11 above)
Comment (ilya-zuyev) [reply test/integration/functional_test.go:367]: done

File: test/integration/functional_test.go [hunk 15]
```diff
@@ -493,7 +513,26 @@ func validateDryRun(ctx context.Context, t *testing.T, profile string) {
 	defer cancel()
 
 	// Too little memory!
-	startArgs := append([]string{"start", "-p", profile, "--dry-run", "--memory", "250MB", "--alsologtostderr"}, StartArgs()...)
+	var startArgs []string
+	orgStartArgs := StartArgs()
+
+	// remove '--memory=xxx' and '--memory xxx' arguments
```
Comment (medyagh) [inline test/integration/functional_test.go:519]: if this change is not related to this PR or not needed for arm64 lets not include it in this PR.

File: test/integration/functional_test.go [hunk 16]
```diff
@@ -493,7 +513,26 @@ func validateDryRun(ctx context.Context, t *testing.T, profile string) {
...
+			continue
+		}
+		if arg == "--memory" {
+			i++ // skip next
+			continue
+		}
+		startArgs = append(startArgs, arg)
+	}
+
+	startArgs = append([]string{"start", "-p", profile, "--dry-run", "--memory", "250MB", "--alsologtostderr"}, startArgs...)
```
Comment (medyagh) [inline test/integration/functional_test.go:535]: we should keep using the original "StartArgs()" helper.

File: test/integration/functional_test.go [hunk 17]
```diff
@@ -783,7 +822,14 @@ func validateServiceCmd(ctx context.Context, t *testing.T, profile string) {
 		}
 	}()
 
-	rr, err := Run(t, exec.CommandContext(ctx, "kubectl", "--context", profile, "create", "deployment", "hello-node", "--image=k8s.gcr.io/echoserver:1.4"))
+	var echoServerArg string
+	if arm64Platform() {
```
Comment (medyagh) [inline test/integration/functional_test.go:826]: and why do we have two different echo server versions ?

if 1.10 is not availble in arm, how about making both to be 1.8

**Review by medyagh — CHANGES_REQUESTED** [review]

File: test/integration/functional_test.go (hunk 17 above)
Comment (ilya-zuyev) [reply test/integration/functional_test.go:808]: Yeah, echoserver is not multiarch and the last available arm version was 1.8
Changed to 1.8 for all archs

File: test/integration/functional_test.go (hunk 16 above)
Comment (ilya-zuyev) [reply test/integration/functional_test.go]: fixed

File: test/integration/functional_test.go (hunk 13 above)
Comment (ilya-zuyev) [reply test/integration/functional_test.go]: done

File: test/integration/functional_test.go (hunk 15 above)
Comment (ilya-zuyev) [reply test/integration/functional_test.go]: done

**Review by medyagh — APPROVED** [review]
